### PR TITLE
Fix cache key generation.

### LINKED ISSUE #16610: Table::get with cache option generates invalid cachekey
### Description

When querying an entity with `Table::get` and cache option activated but without specifying a cacheKey, Cake generates an invalid cache key name if the primary key is a string.

Client's code:
```php
$someStringId = "myStringUid";
$entity = $table
                ->get($someStringId, [
                    'cache' => 'default',
                    'key' => false
                ]);
```

Returns "Cache key `get-connection-name-table-["myStringUid"]` contains invalid characters. You cannot use /, \\, <, >, ?, :, |, *, or " in cache keys."

The primary key is json encoded, therefore the generated cache key contains forbiden characters such as `"` when primary key is a string.

### CakePHP Version

4.3.7

### PHP Version

8.0

### PR DESCRIPTION
**Pull request by ADmad**
Double quotes are not allowed in cache key. Closes #16610

### CODE REVIEW

File: src/ORM/Table.php [hunk 1]
```diff
@@ -1531,7 +1533,7 @@ public function get($primaryKey, array $options = []): EntityInterface
                     'get-%s-%s-%s',
                     $this->getConnection()->configName(),
                     $this->getTable(),
-                    json_encode($primaryKey)
+                    str_replace('"', '', json_encode($primaryKey))
```
Comment (ADmad) [inline src/ORM/Table.php]: This is quite hackish but I wanted to ensure the key doesn't change for integer ids. We should switch to a better approach for 5.x.

File: src/ORM/Table.php [hunk 2]
```diff
@@ -1503,7 +1503,9 @@ public function get($primaryKey, array $options = []): EntityInterface
         foreach ($key as $index => $keyname) {
             $key[$index] = $alias . '.' . $keyname;
         }
-        $primaryKey = (array)$primaryKey;
+        if (!is_array($primaryKey)) {
+            $primaryKey = [$primaryKey];
```
Comment (ADmad) [inline src/ORM/Table.php:1514]: @markstory If the idea to use `json_encode()` was to cater for datetime primary keys then that didn't work either with current code. Casting the primary key as done earlier caused the datetime object to be converted to it's array representation and caused a `InvalidPrimaryKeyException` to be thrown below :slightly_smiling_face:.

**Comment by ADmad** [comment]
Hmm the `FrozenDate` instance gets cast to `2020-07-08` for me locally but to `2020-07-08T00:00:00Z` on GA. Also since the "toString" format is customizable by user there could be more invalid chars :(

**Comment by ADmad** [comment]
The issue occurred only for `FileEngine`, other cache engines wouldn't have problem with the `"` in key. So this change would invalidate the cache for those using string primary keys and not using `FileEngine`.

File: src/ORM/Table.php (hunk 2 above)
Comment (markstory) [reply src/ORM/Table.php:1514]: Well that theory is bogus then. Json was probably easy to use at the time :smile:

**Comment by markstory** [comment]
> So this change would invalidate the cache for those using string primary keys and not using FileEngine.

We can mention the cache key changes in the release notes.

**Comment by garas** [comment]
As alternative, `FileEngine` could allow all symbols by escaping filename with `rawurlencode()`.

**Comment by ADmad** [comment]
@garas I like that idea.

**Comment by ADmad** [comment]
I have updated the PR to just fix the array casting.

**Comment by markstory** [comment]
> As alternative, FileEngine could allow all symbols by escaping filename with rawurlencode().

Should we include this fix in this pull request? The key generation will still fail with FileEngine right now.

**Comment by garas** [comment]
> > As alternative, FileEngine could allow all symbols by escaping filename with rawurlencode().
> 
> Should we include this fix in this pull request? The key generation will still fail with FileEngine right now.

Alternate PR in https://github.com/cakephp/cakephp/pull/16619

**Comment by ADmad** [comment]
> Alternate PR in #16619

Complimentary PR :)

**Pull request merged**